File: MK-52_ESP32/src/Receiver_DATA_F.cpp
//////////////////////////////////////////////////////////
//
//  MK-52 RESURRECT
//  Copyright (c) 2020 Mike Yakimov.  All rights reserved.
//  See main file for the license
//
//////////////////////////////////////////////////////////

#include "Receivers.hpp"
#include "../functions/functions.h"

#define __DEBUG

using namespace MK52_Interpreter;

unsigned long Receiver_DATA_F::init( void *components[]) {
    #ifdef __DEBUG
    Serial.println( "Init DATA_F");
    #endif
    return Receiver::init(components);
}

void Receiver_DATA_F::activate( uint8_t scancode, int8_t parent){
    #ifdef __DEBUG
    Serial.println( "Activating receiver DATA_F");
    #endif
    Receiver::activate( scancode, parent);
    _lcd->updateStatusFMODE( " F ");
    _mode = 1;
    if(!scancode) return;
    tick(scancode);
}

int Receiver_DATA_F::tick( uint8_t scancode){
    int return_value = COMPONENT_RECEIVER_DATA_N;
    switch( scancode){
        // Column 0
        case 2:
            return_value = COMPONENT_RECEIVER_DATA_K;
            break;
        case 3:
            return_value = COMPONENT_RECEIVER_DATA_A;
            break;
        case 4:
            _rpnf->execute(FUNC_TOGGLE_DMOD);
            _lcd->updateStatusDMODE(_rpnf->rpnStack->getDModeName());
            return NO_CHANGE;

        // Column 1 does nothing
        // Column 2 does nothing
        // Column 3 does nothing
        // Column 4 does nothing

        // Column 5
        case 24:
            #ifdef __DEBUG
            Serial.println("Going to AUTO Display");
            #endif
            return_value = COMPONENT_DISPLAY_AUTO;
            break;

        // Column 6
        case 28:
            #ifdef __DEBUG
            Serial.println("Going to PROG Display");
            #endif
            return_value = COMPONENT_DISPLAY_PROG;
            break;

        // Column 7
        case 32:
            break;

        default: // all other buttons do nothing, keeping F-mode
           return NO_CHANGE;
    }
    _mode = 0;
    return return_value;
}

File: MK-52_Simulator/Receivers/Receiver_DATA_F.cs
//////////////////////////////////////////////////////////
//
//  MK-52 RESURRECT
//  Copyright (c) 2020 Mike Yakimov.  All rights reserved.
//  See main file for the license
//
//////////////////////////////////////////////////////////

using System;
using System.Collections.Generic;
using System.Text;
using System.IO;

namespace MK52Simulator
{
    //
    // Implements a receiver for F-modified list
    //
    public class Receiver_DATA_F: Receiver_DATA
    {
        public Receiver_DATA_F(MK52_Host parent)
            : base( parent)
        {
            Moniker = "DATA_F";
        }

        public override void activate(string prevReceiver)
        {
            base.activate(prevReceiver);
            LCD_Manager lm = _parent.getLCD();
            lm.updateStatusFMODE(" F ");
            lm.forcePaint();
        }

        public override byte tick(byte scancode)
        {
            RPN_Functions _rpnf = _parent.getFunctions();
            switch (scancode)
            {
                case 0: // keyboard inactive
                    return 0;

                // Column 0
                case 2:
                    return _rpnf.requestNextReceiver("DATA_K");
                case 3:
                    return _rpnf.requestNextReceiver("DATA_A");
                case 4:
                    //_parent._m_RPN_Stack.toggleAngleMode();
                    //return base.tick(0);
                    return 0;

                // Column 1
                case 5:
                    for( int i=0; i<9; i++)
                        _rpnf.execute(RPN_Functions.FUNC_INCREMENT_MC);
                    break;
                case 6:
                    for (int i = 0; i < 9; i++)
                        _rpnf.execute(RPN_Functions.FUNC_DECREMENT_MC);
                    break;
                //case 7:
                //    TODO: goto to the next entry
                //    break;
                // Column 2 does nothing (for now)
                // Column 3 does nothing (for now)
                // Column 4 does nothing (for now)

                // Column 5
                case 24:
                    return _rpnf.requestNextReceiver("AUTO_N");

                // Column 6
                case 28:
                    return _rpnf.requestNextReceiver("PROG_N");

                // Column 7
                case 32:
                    // Clear F mode
                    break;

                case 33:
                    // Shutdown signal
                    _rpnf.requestNextReceiver("DATA_N");
                    return 33;

                default: // all other buttons do nothing, keeping F-mode
                    return 0;
            }
            return _rpnf.requestNextReceiver("DATA_N");
        }
    }
}
